src/server/acl/user.h

```diff
@@ -33,23 +33,55 @@ class User final {
   struct UpdateRequest {
     std::optional<std::string> password{};
 
-    std::vector<std::pair<Sign, uint32_t>> categories;
-
     std::optional<bool> is_active{};
 
     bool is_hashed{false};
 
+    // Categories and commands
+    using CategoryValueType = std::pair<Sign, uint32_t>;
     // If index s numberic_limits::max() then it's a +all flag
     using CommandsValueType = std::tuple<Sign, size_t /*index*/, uint64_t /*bit*/>;
-    using CommandsUpdateType = std::vector<CommandsValueType>;
-    CommandsUpdateType commands;
+    using UpdateType = std::vector<std::variant<CategoryValueType, CommandsValueType>>;
+    UpdateType updates;
 
     // keys
     std::vector<UpdateKey> keys;
     bool reset_all_keys{false};
     bool allow_all_keys{false};
   };
 
+  struct CategoryChange {
+    uint32_t category;
+
+    // Customization point to make it hashable with absl containers
+    template <typename H> friend H AbslHashValue(H h, const CategoryChange& c) {
+      return H::combine(std::move(h), c.category);
+    }
+
+    friend bool operator==(CategoryChange c1, CategoryChange c2) {
+      return c1.category == c2.category;
+    }
+  };
+
+  struct CommandChange {
+    size_t family = 0;
+    uint64_t bit_index = 0;
+
+    // Customization point to make it hashable with absl containers
+    template <typename H> friend H AbslHashValue(H h, const CommandChange& c) {
+      return H::combine(std::move(h), c.family + c.bit_index);
+    }
+
+    friend bool operator==(CommandChange c1, CommandChange c2) {
+      return (c1.family == c2.family) && (c1.bit_index == c2.bit_index);
+    }
+  };
```

Comment:
fun fact alias CommandChange = pair<> is also hashable, assingnable and comparable 🙂 I doubt you'll ever "accidentally" assign a regular pair to it 🙃 

I don't see it being used in a variant